### PR TITLE
Allow reification of third-party libraries, closes #14

### LINKED ISSUE #14: Allow overriding reification of the standard and third party libraries.
The library should be able to reify constructs from the standard and third party libraries. A plausible solution is to replace the symbols of third-party types with configurable ones and re-typecheck the code. For example, in: 
```
val x = 1
x + 1
```
type of `x` could be replaced with `dsl.scala.Int` where  `dsl.scala.Int` would be:
```
package dsl.scala

class Int {
  @reifyAs(IntPlus)
  def +(x: Int): Int = ???
}
```

### PR DESCRIPTION
**Pull request by olafurpg**
We configure the overriding with a `typeMap: Map[Type, Type]`.  We
perform the type overriding as an integral part of the ReifyAsEmbedding
phase, more precisely in the `body` method.

The benefit of this approach is that we don't have to type-rewrite any
nodes in the block. This means the reification of standard and
third-party types will only trigger if the DSL author provides an
implementation of the invoked method in the corresponding type in
`typeMap`. If no matching method (see bottom comment) with a `reifyAs`
annotation is found in the corresponding type the transformation falls
back to the general error message, i.e. 'method X is not supported in
DSL'.

The limitation of this approach is mainly that we cannot override a type
if the method already contains a `reifyAs` annotation. We accept this
limitation since (presumably) the standard and third-party libraries
won't have the annotations in the first place.

This commit does not provide a solution for method overloading.

### CODE REVIEW

**Comment by olafurpg** [comment]
Review @vjovanov

File: tests/src/test/scala/ch/epfl/directembedding/test/TypeOverridingSpec.scala [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import org.scalatest.{ FlatSpec, ShouldMatchers }
+
+class TypeOverridingSpec extends FlatSpec with ShouldMatchers with ExampleTester {
+
+  "dsl" should "allow custom overriding of ints" in {
+    runTest(
+      dsl {
+        val i = 1
+        i + 1
+      }) should be(IntPlus(Infix_ValDef(1), 1))
```
Comment (olafurpg) [inline tests/src/test/scala/ch/epfl/directembedding/test/TypeOverridingSpec.scala]: Since we virtualize val defs we expect a `Infix_ValDef` on the left hand side.

File: tests/src/test/scala/ch/epfl/directembedding/test/TypeOverridingSpec.scala [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  // type of `s` is String and the result is of type Int
+  "dsl" should "allow custom overriding of strings" in {
+    runTest(
+      dsl {
+        val s = "foobar"
+        s.length
+      }) should be(StringLength(Infix_ValDef("foobar")))
+  }
+
+  // TODO: Add test cases for all branches in ReifyAsEmbedding
```
Comment (olafurpg) [inline tests/src/test/scala/ch/epfl/directembedding/test/TypeOverridingSpec.scala]: Remove? There are a lot of untested cases but I thought it might be good to do that through another PL.

**Comment by olafurpg** [comment]
I added more tests and fixed one bug with `Typed` blocks.

File: directembedding/src/main/scala/ch/epfl/directembedding/transformers/ReifyAsEmbedding.scala [hunk 3]
```diff
@@ -84,42 +80,93 @@ trait ReifyAsEmbedding extends DirectEmbeddingModule with DirectEmbeddingUtils {
...
       val result = tree match {
+        // f(a)(b) => f(a, b)
         case a @ Apply(Apply(_, _), _) =>
-          logIndented(s"Apply(Apply)", indent, logLevel)
+          logIndented(s"Apply(Apply)", indent)
           transform(uncurry(a, Nil))
 
+        // (expr: typ) => expr
+        case Typed(expr, typ) =>
+          transform(expr)
```
Comment (olafurpg) [inline directembedding/src/main/scala/ch/epfl/directembedding/transformers/ReifyAsEmbedding.scala]: Is this sufficient to solve typed blocks or do we need to use `TypeTransformation` from Yin-Yang?

File: tests/src/test/scala/ch/epfl/directembedding/test/ErrorsSpec.scala [hunk 4]
```diff
@@ -9,11 +9,20 @@ class ErrorsSpec extends FlatSpec with ShouldMatchers with ExampleTester {
...
       """
-        runTest(
-          dsl {
-            ObjectExample.missingAnnotation
-        }) should be(???)
-      """, "method missingAnnotation is not supported in example.dsl")
+        dsl {
+          ObjectExample.missingAnnotation
+        }
+      """, "method missingAnnotation on object ObjectExample is not supported in example.dsl")
```
Comment (olafurpg) [inline tests/src/test/scala/ch/epfl/directembedding/test/ErrorsSpec.scala]: The position in the error message is now also on the invoked symbol instead of the outer `dsl` block, a big improvement.

File: directembedding/src/main/scala/ch/epfl/directembedding/Utils.scala [hunk 5]
```diff
@@ -13,8 +13,12 @@ trait MacroModule {
 }
 
 trait DirectEmbeddingModule extends MacroModule {
-  import c.universe.Tree
+  import c.universe._
   val dslName: String
+
+  // We use Strings as keys to resolve aliased types
+  val typeMap: Map[String, Type]
```
Comment (vjovanov) [inline directembedding/src/main/scala/ch/epfl/directembedding/Utils.scala]: This is good for now! Let's see what will be the best thing in practice.

File: directembedding/src/main/scala/ch/epfl/directembedding/Utils.scala [hunk 6]
```diff
@@ -87,3 +93,16 @@ trait DirectEmbeddingUtils extends MacroModule with TransformationUtils {
   }
 
 }
+
+trait TypeHelper extends MacroModule {
+  import c.universe._
+
+  private def name(s: Symbol): String = s.name.toString.trim
+
+  def equalSymbols(a: Symbol)(b: Symbol): Boolean =
```
Comment (vjovanov) [inline directembedding/src/main/scala/ch/epfl/directembedding/Utils.scala]: Does this work in practice? Shouldn't we check this in a bit more involved way.

File: directembedding/src/main/scala/ch/epfl/directembedding/transformers/ReifyAsEmbedding.scala (hunk 3 above)
Comment (vjovanov) [inline directembedding/src/main/scala/ch/epfl/directembedding/transformers/ReifyAsEmbedding.scala]: I think we will need it for now. This can go into the separate pull request.

File: tests/src/test/scala/ch/epfl/directembedding/test/ErrorsSpec.scala (hunk 4 above)
Comment (vjovanov) [inline tests/src/test/scala/ch/epfl/directembedding/test/ErrorsSpec.scala]: Awesome!

**Comment by vjovanov** [comment]
Otherwise, LGTM.

File: directembedding/src/main/scala/ch/epfl/directembedding/Utils.scala (hunk 6 above)
Comment (olafurpg) [inline directembedding/src/main/scala/ch/epfl/directembedding/Utils.scala]: This has worked for all the tests so far, but I agree there must be a better way.

File: directembedding/src/main/scala/ch/epfl/directembedding/transformers/ReifyAsEmbedding.scala (hunk 3 above)
Comment (olafurpg) [inline directembedding/src/main/scala/ch/epfl/directembedding/transformers/ReifyAsEmbedding.scala]: I removed this case along with the corresponding test case.

**Comment by olafurpg** [comment]
Feel free to merge.

File: directembedding/src/main/scala/ch/epfl/directembedding/Utils.scala (hunk 6 above)
Comment (vjovanov) [inline directembedding/src/main/scala/ch/epfl/directembedding/Utils.scala]: Let's at least do a type of the symbol and if it is a method number of arguments. I think we could also add tests for different kind of weird situations. Also, I think it will be fun to see different kinds of symbols and how to compare them.

Comment (vjovanov) [inline directembedding/src/main/scala/ch/epfl/directembedding/Utils.scala]: Also, if this one does not inspire you I am well motivated to try it out.

Comment (olafurpg) [inline directembedding/src/main/scala/ch/epfl/directembedding/Utils.scala]: Feel free to take on this one, I won't be able to look into it until Thursday.

I don't know if it's necessary, but it may help to change the signature to include the trees.

Comment (vjovanov) [inline directembedding/src/main/scala/ch/epfl/directembedding/Utils.scala]: Cool, I will look into it.
